src/docs/tool-libfuzzer.dox

```diff
@@ -0,0 +1,210 @@
+@endcode
+
+Each fuzz target will produce a few outputs:
+
+- \c fuzz-N.log: The LibFuzzer log for worker N.
+
+- \c WT_TEST_<pid>: The home directory for a given worker process.
+
+- \c crash-<input-hash>: If an error occurs, you may see a file like this containing the input that
+crashed the target.
+
+- \c WT_TEST_<pid>.profraw: If you're running with Clang coverage (more on this later), you'll see
+files containing profiling data for a given worker. These will be used by \c fuzz_coverage.
+
+### Corpus
+
+LibFuzzer is a coverage based fuzzer meaning that it notices when a particular input hits a new code
+path and adds it to a "corpus" of interesting data inputs. It then uses existing data in the corpus
+to mutate and come up with new inputs.
+
+While LibFuzzer will automatically add to its corpus when it finds an interesting input, some fuzz
+targets (especially those that expect data in a certain format) require a corpus to start things off
+in order to be effective. The target \c fuzz_config is one example of this as it expects its data
+sliced with a separator so the fuzzer needs some examples to guide it. The corpus is supplied as the
+first positional argument to both \c fuzz_run.sh and the underlying fuzz target itself. For example:
+
+@code
+$ cd build_posix/test/fuzz/
+$ bash ../../../test/fuzz/fuzz_run.sh ./fuzz_config ../../../test/fuzz/config/corpus/
+@endcode
+
+# Implementing an LLVM LibFuzzer target
+
+## Overview
+
+So you have functionality that you'd like to test against LLVM LibFuzzer and would like to add a new
+target to the system.
+
+We do this by implementing a function with the following prototype:
+
+@code
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+@endcode
+
+When we supply the \c -fsanitize=fuzzer flag, Clang will use its own \c main function and repeatedly
+call your \c LLVMFuzzerTestOneInput with various inputs. There is a lot of information online about
+best practices when writing fuzzing targets but to summarize, the requirements are much like those
+for writing unit tests: they should be fast, deterministic and stateless (as possible). The
+[LLVM LibFuzzer reference page](https://llvm.org/docs/LibFuzzer.html) is a good place to start if
+you'd like to learn more.
+
+## Fuzz Utilities
+
+WiredTiger has a small fuzz utilities library containing common functionality required for writing
+fuzz targets. Most of the important functions here are about manipulating fuzz input into something
+that your target can use with a WiredTiger API.
+
+### Slicing inputs
+
+If the function that you're trying to exercise can accept a binary blob of data, then your target
+will be straightforward as it'll more or less just pass the input to the function under test. But
+for functions that require more structured input, this can pose a challenge. As an example, the
+\c fuzz_config target requires two inputs that we want to be variable length: a configuration string
```

Comment:
I am wondering if some of this documentation would be helped by an example. You can pull in snippets from working code into the documentation to show certain specific things. In the source code you put a beginning/ending comment to the snippet with `/*! [Some text] */`. You can look at `examples/c/ex_stat.c` for marking the code, and `src/docs/data-sources.dox` for how to pull it into the documentation.

Reply:
1a37f01

Thanks, that's really nifty.